strawberryfields/backends/states.py

```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
 
         return mean, var
 
+    def poly_quad_expectation(self, A, d=None, k=0, phi=0, **kwargs):
+        if A is None:
+            A = np.zeros([2*self._modes, 2*self._modes])
+
+        if A.shape != (2*self._modes, 2*self._modes):
+            raise ValueError("Matrix of quadratic coefficients A must be of size 2Nx2N.")
+
+        if np.any(A.T != A):
+            raise ValueError("Matrix of quadratic coefficients A must be symmetric.")
+
+        if d is None:
+            linear_coeff = np.zeros([2*self._modes])
+        else:
+            linear_coeff = d.copy()
+            linear_coeff[self._modes:] = -d[self._modes:]
+
+        if linear_coeff.shape != (2*self._modes,):
+            raise ValueError("Vector of linear coefficients d must be of length 2N.")
+
+        # expand the cutoff dimension in approximating the x and p
+        # operators in the Fock basis, to reduce numerical inaccuracy.
+        worksize = kwargs.get('worksize', 5)
+        dim = self._cutoff + worksize
+
+        # construct the x and p operators
+        a = np.diag(np.sqrt(np.arange(1, dim)), 1)
+        x = np.sqrt(self._hbar/2) * (a + a.T)
+        p = -1j * np.sqrt(self._hbar/2) * (a - a.T)
+
+        if phi != 0:
+            # rotate the quadrature operators
+            xold = x.copy()
+            pold = p.copy()
```

Comment:
Only because we have simultaneous linear equations in the next two lines, using the same variable name. Since `x` and `p` are NumPy arrays, using temporary variables does not work, as it provides a view to the same array.

Reply:
What if we just use different names for the transformed operators? Then we wouldn't need to copy, right?

```
x_ = np.sqrt(...)
p_ = -1j * ...

if phi != 0:
    x = np.cos(phi) * x - np.sin(phi) * p
    p = np.sin(phi) * x + np.sin(phi) * p
else:
    x = x_
    p = p_